src/praisonai-agents/praisonaiagents/llm/llm.py

```
@@ -432,13 +437,163 @@ def _supports_xml_tool_format(self) -> bool:
+
+                    logging.warning(
+                        f"Rate limit hit (attempt {attempt + 1}/{self._max_retries + 1}), "
+                        f"waiting {retry_delay:.1f}s before retry..."
+                    )
+
+                    if self._rate_limiter is not None:
+                        self._rate_limiter.wait_for_retry(retry_delay)
+                    else:
+                        time.sleep(retry_delay)
+                else:
+                    logging.error(f"Rate limit exceeded after {self._max_retries + 1} attempts")
```

Ruff (0.14.10)

Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)